skbio/diversity/_driver.py

```diff
@@ -0,0 +1,304 @@
+    counts : 1D or 2D array_like of ints or floats
+        Vector or matrix containing count/abundance data. If a matrix, each row
+        should contain counts of OTUs in a given sample.
+    ids : iterable of strs, optional
+        Identifiers for each sample in ``counts``.
+    validate: bool, optional
+        If `False`, validation of the input won't be performed. This step can
+        be slow, so if validation is run elsewhere it can be disabled here.
+        However, invalid input data can lead to invalid results or error
+        messages that are hard to interpret, so this step should not be
+        bypassed if you're not certain that your input data are valid. See
+        Notes for the description of what validation entails so you can
+        determine if you can safely disable validation.
+    kwargs : kwargs, optional
+        Metric-specific parameters.
+
+    Returns
+    -------
+    pd.Series
+        Values of ``metric`` for all vectors provided in ``counts``. The index
+        will be ``ids``, if provided.
+
+    Raises
+    ------
+    ValueError, MissingNodeError, DuplicateNodeError
+        If validation fails (see description of validation in Notes). Exact
+        error will depend on what was invalid.
+    TypeError
+        If invalid method-specific parameters are provided.
+
+    See Also
+    --------
+    skbio.diversity.beta_diversity
+
+    Notes
+    -----
+    The value that you provide for ``metric`` can be either a string (e.g.,
+    ``"faith_pd"``) or a function (e.g., ``skbio.diversity.alpha.faith_pd``).
+    The metric should generally be passed as a string, as this often uses an
+    optimized version of the metric. For example, passing  ``"faith_pd"`` (a
+    string) will be tens of times faster than passing the function
+    ``skbio.diversity.alpha.faith_pd``. The latter may be faster if computing
+    alpha diversity for only one or a few samples, but in these cases the
+    difference in runtime is negligible, so it's safer to just err on the side
+    of passing ``metric`` as a string.
+
+    Validation of input data confirms the following:
+     * ``counts`` data can be safely cast to integers
+     * there are no negative values in ``counts``
+     * ``counts`` has the correct number of dimensions
+     * if ``counts`` is 2-D, all vectors are of equal length
+     * the correct number of ``ids`` is provided (if any are provided)
+    And for phylogenetic diversity metrics:
+     * ``otu_ids`` does not contain duplicate values
+     * the length of each ``counts`` vector is equal to ``len(otu_ids)``
+     * ``tree`` is rooted
+     * ``tree`` has more than one node
+     * all nodes in ``tree`` except for the root node have branch lengths
+     * all tip names in ``tree`` are unique
+     * all ``otu_ids`` correspond to tip names in ``tree``
+
+    """
+    metric_map = _get_alpha_diversity_metric_map()
+
+    if validate:
+        counts = _validate_counts_matrix(counts, ids=ids)
```

Comment:
`_validate_counts_matrix` is validating *and* casting input into the correct type (mainly transforming `array_like` into `ndarray`). Thus, passing an `array_like` with `validate=False` won't work (a bug).

I think we need the input casting and validation separated but that should probably be tackled in #1186. @gregcaporaso do you agree? If so can you add a note to that issue so this isn't forgotten?